python/lsst/daf/butler/remote_butler/_http_connection.py

```diff
@@ -0,0 +1,226 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("RemoteButlerHttpConnection", "parse_model")
+
+from collections.abc import Mapping
+from typing import TypeVar
+from uuid import uuid4
+
+import httpx
+from lsst.daf.butler import __version__
+from pydantic import BaseModel, ValidationError
+
+from .._exceptions import create_butler_user_error
+from ._authentication import get_authentication_headers
+from .server_models import CLIENT_REQUEST_ID_HEADER_NAME, ERROR_STATUS_CODE, ErrorResponseModel
+
+_AnyPydanticModel = TypeVar("_AnyPydanticModel", bound=BaseModel)
+"""Generic type variable that accepts any Pydantic model class."""
+
+
+class RemoteButlerHttpConnection:
+    """HTTP connection to a Butler server.
+
+    Parameters
+    ----------
+    http_client : `httpx.Client`
+        HTTP connection pool we will use to connect to the server.
+    server_url : `str`
+        URL of the Butler server we will connect to.
+    access_token : `str`
+        Rubin Science Platform Gafaelfawr access token that will be used to
+        authenticate with the server.
+    """
+
+    def __init__(self, http_client: httpx.Client, server_url: str, access_token: str) -> None:
+        self._client = http_client
+        self.server_url = server_url
+        self._access_token = access_token
+
+        auth_headers = get_authentication_headers(access_token)
+        headers = {"user-agent": f"RemoteButler/{__version__}"}
+
+        self._headers = auth_headers | headers
+
+    def post(self, path: str, model: BaseModel) -> httpx.Response:
+        """Send a POST request to the Butler server.
+
+        Parameters
+        ----------
+        path : `str`
+            A relative path to an endpoint.
+        model : ``BaseModel``
+            Pydantic model containing the request body to be sent to the
+            server.
+
+        Returns
+        -------
+        response: `~httpx.Response`
+            The response from the server.
+
+        Raises
+        ------
+        ButlerUserError
+            If the server explicitly returned a user-facing error response.
+        ButlerServerError
+            If there is an issue communicating with the server.
+        """
+        json = model.model_dump_json(exclude_unset=True).encode("utf-8")
+        return self._send_request(
+            "POST",
+            path,
+            content=json,
+            headers={"content-type": "application/json"},
+        )
+
+    def get(self, path: str, params: Mapping[str, str | bool] | None = None) -> httpx.Response:
+        """Send a GET request to the Butler server.
+
+        Parameters
+        ----------
+        path : `str`
+            A relative path to an endpoint.
+        params : `Mapping` [ `str` , `str` | `bool` ]
```

Comment:
```suggestion
        params : `~collections.abc.Mapping` [ `str` , `str` | `bool` ]
```
Sphinx requires this to be fully specified.